grpc/src/tx.rs

```diff
@@ -0,0 +1,602 @@
+            let tx = sign_tx(tx.clone(), 0, 1);
+            let gas_info = self.client.simulate(tx.encode_to_vec()).await?;
+            (gas_info.gas_used as f64 * DEFAULT_GAS_MULTIPLIER) as u64
+        };
+
+        let gas_price = cfg.gas_price.unwrap_or(self.gas_price());
+        let fee = (gas_limit as f64 * gas_price).ceil();
+        let tx = sign_tx(tx, gas_limit, fee as u64);
+
+        self.broadcast_tx_with_account(tx.encode_to_vec(), account, gas_limit)
+            .await
+    }
+
+    async fn sign_and_broadcast_blobs(
+        &self,
+        blobs: Vec<Blob>,
+        cfg: TxConfig,
+    ) -> Result<(Hash, u64)> {
+        // lock the account; tx signing and broadcast must be atomic
+        // because node requires all transactions to be sequenced by account.sequence
+        let account = self.account.lock().await;
+
+        let pfb = MsgPayForBlobs::new(&blobs, account.address.clone())?;
+        let pfb = RawTxBody {
+            messages: vec![into_any(RawMsgPayForBlobs::from(pfb))],
+            ..RawTxBody::default()
+        };
+
+        let gas_limit = cfg
+            .gas_limit
+            .unwrap_or_else(|| estimate_gas(&blobs, self.app_version, DEFAULT_GAS_MULTIPLIER));
+        let gas_price = cfg.gas_price.unwrap_or(self.gas_price());
+        let fee = (gas_limit as f64 * gas_price).ceil() as u64;
+        let tx = sign_tx(
+            pfb,
+            self.chain_id.clone(),
+            &account,
+            &self.pubkey,
+            &self.signer,
+            gas_limit,
+            fee,
+        );
+
+        let blobs = blobs.into_iter().map(Into::into).collect();
+        let blob_tx = RawBlobTx {
+            tx: tx.encode_to_vec(),
+            blobs,
+            type_id: BLOB_TX_TYPE_ID.to_string(),
+        };
+
+        self.broadcast_tx_with_account(blob_tx.encode_to_vec(), account, gas_limit)
+            .await
+    }
+
+    async fn broadcast_tx_with_account(
+        &self,
+        tx: Vec<u8>,
+        mut account: MutexGuard<'_, Account>,
+        gas_limit: u64,
+    ) -> Result<(Hash, u64)> {
+        let resp = self.client.broadcast_tx(tx, BroadcastMode::Sync).await?;
+
+        if resp.code != ErrorCode::Success {
+            return Err(Error::TxBroadcastFailed(
+                resp.txhash,
+                resp.code,
+                resp.raw_log,
+                gas_limit,
+            ));
+        }
+
+        let tx_sequence = account.sequence;
+        account.sequence += 1;
+
+        Ok((resp.txhash, tx_sequence))
+    }
+
+    async fn confirm_tx(&self, hash: Hash, sequence: u64) -> Result<TxInfo> {
+        let mut interval = Interval::new(Duration::from_millis(500)).await;
+
+        loop {
+            let tx_status = self.client.tx_status(hash).await?;
+            match tx_status.status {
+                TxStatus::Pending => interval.tick().await,
+                TxStatus::Unknown => return Err(Error::TxNotFound(hash)),
```

Comment:
Is it possible to get this even if `broadcast_tx` succeeded? If yes, we need to revert sequence

Reply:
done